### ISSUE TITLE
When retry, queue not processed in order

### ISSUE THREAD

**Issue by akujbida** [diamondio/better-queue#96]
No matter which combination of options I try, if there's ever a retry, the next item is processed before the retry.

Is it possible to convince the queue order to be maintained when there's a retry? I'm sure there must be, since that is largely what queue means.

Example:
```
const Queue = require('better-queue');

function qProcessOutputHandler(err, result) {
	if (err) {
		global.retryCount++;
		console.error(`retry count ${global.retryCount}, process completed but with an error: `, err);
	} else {
		console.log(`Item is done, retry count is ${global.retryCount}, output is: `, result);
	}
}

global.retryCount = 0;

function processWithRetry(item, cb) {
	console.log('Processing item:', item);
	// a small delay to simulate some work
	// For task item 2, we'll have the timeout greater than the queue timeout to demonstrate
	// that the queue will retry the item
	const t = item === 2 ? 400 : 100;
	setTimeout(() => {
		if ((item == 2) && (global.retryCount++ < 5)) {
			if (global.retryCount == 3) { // pretend we finally succeeded
				console.log("pretending success, item: ", item, ", count: ", global.retryCount);
				cb(null, {item, info: 'job completed with a struggle, this is the result object'});
			}
			else {
				console.log("pretending something is wrong, retrying item: ", item);
				cb(new Error(`pretending something is wrong, retrying item. count: ${global.retryCount}, item:` + item, '.'), item);
			}
		}
		else {
			cb(null, {item, info: 'job completed ok, this is the result object'});
		}
	}, t);
}

const q = new Queue(processWithRetry, {concurrent: 1, maxRetries: 10, retryDelay: 0,  maxTimeout: 1000});

// add some items to the queue
q.push(1, qProcessOutputHandler);
q.push(2, qProcessOutputHandler);
q.push(3, qProcessOutputHandler);

//Output:
Processing item: 1
Item is done, retry count is 0, we'll reset count. output is:  { item: 1, info: 'job completed ok, this is the result object' }
Processing item: 2
pretending something is wrong, retrying item:  2
Processing item: 3
Item is done, retry count is 1, we'll reset count. output is:  { item: 3, info: 'job completed ok, this is the result object' }
Processing item: 2
pretending something is wrong, retrying item:  2
Processing item: 2
pretending success, item:  2 , count:  3
Item is done, retry count is 3, we'll reset count. output is:  { item: 2,  info: 'job completed with a struggle, this is the result object'}
```

**Comment by leanderlee** [comment]
I believe, by default, failed tasks that are re-tried are re-queued, but there's no option to ensure it queues as the next task like you're mentioning. You could probably accomplish this with the priority system, or maybe we can add an option for this.

Alternatively, if you want to retry immediately, why not do so in the process function? Perhaps it should not fail the task until it has tried enough times.

Feel free to submit a PR for that change!

**Comment by akujbida** [comment]
> I believe, by default, failed tasks ...

Thanks very much for the confirmation (and Sunday reply!) Yes, certainly I've looked at putting retries into the process (did that in Perl years ago) and this probably would have been quicker than all the searching and testing I did with existing libs. It seems so obvious a need to me that I was certain it was easily available and better written than I could do. This is for serialport, BTW. Pretty standard requirement in industrial platforms but strangely not available in any serial port module either! I've manually written a childishly simply queue to replace better-queue, though I'm sure I'll regret it :-)

Thanks again for the info!
